Documentation/bootblock-protection.md

```diff
@@ -0,0 +1,193 @@
+chip from writing using `flashrom` and its support for manipulating SPI write
+protection (WP). This kind of protection requires changing connection of WP
+pin of a chip to prevent any attempt of disabling the protection by software
+alone. To not be confused with protection by flash controller of your
+motherboard (PCH protection).
+
+## `flashrom` version
+
+At the time of writing (5 August 2022) there hasn't been a `flashrom` release
+that includes WP manipulation facilities. You might have to build one from
+scratch (assuming you've already installed build dependencies):
+
+```
+git clone --depth 1 https://github.com/flashrom/flashrom
+cd flashrom
+make
+# flashrom executable will appear in current directory
+```
+
+## Chip support by `flashrom`
+
+There is a great variety of chips with some not supporting write protection at
+all and others doing it in their own peculiar way of which `flashrom` has no
+idea. So the first thing to do is to make sure that `flashrom` knows how WP
+works in your particular chip. Run a command like (adjust this and similar
+commands below if you're not using internal programmer or need other options):
+
+```
+flashrom --programmer internal --wp-status
+```
+
+Seeing this output line would mean that `flashrom` doesn't know how to use WP
+feature of the chip you have:
+
+```
+Failed to get WP status: WP operations are not implemented for this chip
+```
+
+Otherwise the output might contain something similar to this:
+
+```
+Protection range: start=0x00000000 length=0x00000000 (none)
+Protection mode: disabled
+```
+
+If so, you can continue with the rest of the instructions.
+
+## Collecting information about the range
+
+You need to know where the area you want to protect starts and ends. The example
+below will assume you're trying to protect bootblock stored in CBFS at the end
+of some coreboot firmware. In other cases it might be a separate file which is
+put at the beginning of a chip. You need to have an idea of what you're doing
+here or have some reliable instructions to follow.
+
+In this case `cbfstool` can be used to list information about bootblock like
+this:
+
+```
+$ cbfstool rom print | sed -n '2p; /bootblock/p'
+Name                           Offset     Type           Size   Comp
+bootblock                      0x3ef100   bootblock       36544 none
+```
+
+However, the offset is relative to the start of CBFS region, so we also need to
+find out offset of CBFS:
+
+```
+$ cbfstool rom layout | grep CBFS
+'COREBOOT' (CBFS, size 4161536, offset 12615680)
+```
+
+Now we can calculate:
+
+ * start offset (CBFS offset + bootblock offset): \
+   `12615680 + 0x3ef100 = 0xff7100` \
+   (`printf "%#x\n" $(( 12615680 + 0x3ef100 ))`)
+ * end offset (start offset + bootblock size - 1): \
+   `0xff7100 + 36544 - 1 = 0xffffbf` \
+   (`printf "%#x\n" $(( 0xff7100 + 36544 - 1 ))`)
+
+Thus we need to write-protect the smallest area that covers the range from
+`0xff7100` to `0xffffbf` (both bounds are inclusive).
```

Comment:
This is not correct, bootblock on x86 (except newest AMD boards that start from RAM) has to end at 0xffffffff. Offset from `cbfstool print` points to some kind of CBFS header that is not included in size reported by that same command. I haven't seen any other size of that metadata than 0x40, but I don't know if this is always true.

Reply:
Yeah, that offset looked weird, but I forgot to check it. Doesn't look like the size is fixed ("offset" here is a field of component header):
```
/** This is a component header - every entry in the CBFS
    will have this header.

    This is how the component is arranged in the ROM:

    --------------   <- 0
    component header
    --------------   <- sizeof(struct component)
    component name
    --------------   <- offset
    data
    ...
    --------------   <- offset + len
*/
```
It's just that `CBFS_ALIGNMENT` is `64` which is why is most cases data will be offset by 64 bytes.

***

Looks like master CBFS header offset is still there on latest `master` (`src/arch/x86/bootblock.ld`) for the case of bootblock in CBFS:
```ld
	. = 0xfffffff0;
	_X86_RESET_VECTOR = .;
	.reset . : {
		*(.reset);
		. = _X86_RESET_VECTOR_FILLING;
		BYTE(0);
	}
	. = 0xfffffffc;
	.header_pointer . : {
		KEEP(*(.header_pointer));
	}
	_X86_RESET_VECTOR_FILLING = 15 - SIZEOF(.header_pointer);
	_ebootblock = .;
```
So it won't work if CBFS size changes. Size of image in 4.17 is 4 MiB vs. 255 KiB for 4.13, I didn't expect bootblock to contain that CBFS offset hard-coded and didn't try to make the size equal.

Reply:
Regarding master header offset, I know it is still there but I haven't seen it being used in current code. Then again, I didn't dig too deep.

Reply:
Even when ROM sizes match, offsets between bootblocks in 4.13 and 4.17 don't (12 byte difference).